crates/trigger-http2/src/server.rs

```diff
@@ -322,11 +335,30 @@ impl HttpServer {
 
 /// The incoming request's scheme and authority
 ///
-/// The incoming request's URI is relative to the server, so we need to set the scheme and authority
-fn set_req_uri(req: &mut Request<Body>, scheme: Scheme, addr: SocketAddr) -> anyhow::Result<()> {
+/// The incoming request's URI is relative to the server, so we need to set the scheme and authority.
+/// The `Host` header is used to set the authority. This function will error if no `Host` header is
+/// present or if it is not parsable as an `Authority`.
+fn set_req_uri(req: &mut Request<Body>, scheme: Scheme) -> anyhow::Result<()> {
     let uri = req.uri().clone();
     let mut parts = uri.into_parts();
-    let authority = format!("{}:{}", addr.ip(), addr.port()).parse().unwrap();
+    let headers = req.headers();
+    let host_header = headers
+        .get(http::header::HOST)
+        .context("missing 'Host' header")?
+        .to_str()
+        .context("'Host' header is not valid UTF-8")?;
+    let authority = host_header
+        .parse()
+        .context("'Host' header contains an invalid authority")?;
+    // Ensure that if `req.authority` is set, it matches what was in the `Host` header
+    // https://github.com/hyperium/hyper/issues/1612
+    if let Some(a) = parts.authority.as_ref() {
+        if a != &authority {
+            return Err(anyhow::anyhow!(
+                "authority in 'Host' header does not match authority in URI"
+            ));
+        }
+    }
```

Comment:
Sorry, I wasn't very clear. I believe it is valid to have just the `host` header, just the uri `authority`, or both set, where in the "both" case they must match.

Egads this seems complicated: https://www.rfc-editor.org/rfc/rfc9113.html#section-8.3.1-2.3.3